modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp

```diff
@@ -0,0 +1,75 @@
+#include "streaming/onevpl/accelerators/surface/surface_pool.hpp"
+#include "streaming/onevpl/accelerators/surface/surface.hpp"
+#include "logger.hpp"
+
+namespace cv {
+namespace gapi {
+namespace wip {
+
+void CachedPool::reserve(size_t size) {
+    surfaces.reserve(size);
+}
+
+size_t CachedPool::size() const {
+    return surfaces.size();
+}
+
+void CachedPool::clear() {
+    surfaces.clear();
+    next_free_it = surfaces.begin();
+    cache.clear();
+}
+
+void CachedPool::push_back(surface_ptr_t &&surf) {
+    cache.insert(std::make_pair(surf->get_handle(), surf));
+    surfaces.push_back(std::move(surf));
+    next_free_it = surfaces.begin();
+}
+
+CachedPool::surface_ptr_t CachedPool::find_free() {
+
+        auto it =
+            std::find_if(next_free_it, surfaces.end(),
+                        [](const surface_ptr_t& val) {
+                GAPI_DbgAssert(val && "Pool contains empty surface");
+                return !val->get_locks_count();
+            });
+
+        // Limitation realloc pool might be a future extension
+        if (it == surfaces.end()) {
+            it =
+                std::find_if(surfaces.begin(), next_free_it,
+                            [](const surface_ptr_t& val) {
+                    GAPI_DbgAssert(val && "Pool contains empty surface");
+                    return !val->get_locks_count();
+                });
+            if (it == next_free_it) {
+                std::stringstream ss;
+                ss << "cannot get free surface from pool, size: " << surfaces.size();
+                const std::string& str = ss.str();
+                GAPI_LOG_WARNING(nullptr, str);
+                throw std::runtime_error(std::string(__FUNCTION__) + " - " + str);
+            }
+        }
+
+        next_free_it = it;
+        ++next_free_it;
```

Comment:
Could you provide some example/link how it works? Is it supposed to be faster than a `vector`?

Reply:
this code will turn up at further PR. The key idea is:
when we tries to get `free surface` we checks existing surface pool vector by getting the first surface with LockCounter == 0 and returns a first one. Naive implementation start searching surface from beginning of array every time. But this implementation remembers last requested surface  iterator + 1 as free surface and merely starts searching free from this position which is 100% free the next time. 
In overall, it is lightweight implementation of single-threaded circular buffer